### PR TITLE
Reworks stop, drop, roll into a gradual, interruptable thing, that repeats until extinguished

### PR DESCRIPTION
**Pull request by MrMelbert**
## About The Pull Request

Related: #78017 

Stop drop and roll is no longer instant -5 fire stacks -> stun -> wait. 

Now, when you stop drop and roll, every time you roll you will lose 1 firestack. 

A roll is triggered every 0.8 seconds. Moving, getting up, or becoming incapacitated / stunned will stop you from rolling. 
_(This number puts it roughly equivalent to its current rate.)_

While rolling, your hands are blocked (you cannot use items, hold things, etc.)
Additionally, you will roll until all firestacks are cleared. 

## Why It's Good For The Game

Getting stunned for 6 seconds because you decide to stop and roll is a little silly. Reasonably you could stop rolling and get back up should the need arise, such as "oh god there's more fire I gotta relocate". 

By changing it to a gradual thing, it makes it a bit more reasonable and fair.
- New players who immediately slam "STOP DROP ROLL" because the alert on their screen tells them to are no longer helpless for 6 whole seconds
- People who hit the resist key, intending to interact with something else (such as a bola) are no longer stuck rolling when they did not want to 

## Changelog

:cl: Melbert
balance: Stop, drop, and roll no longer instantly clears 5 fire stacks off of you - Instead, it will clear 1 fire stack off of you every time you roll, with a roll every 0.8 seconds. 
balance: Stop, drop, and roll no longer stuns you for 6 seconds. Instead, it will knock you to the floor while you are rolling. Moving around or getting up will cancel the roll, and you cannot use items while rolling around.
balance: Stop, drop, and roll will now repeat until the fire is put out or you get up. 
/:cl:

### CODE REVIEW

**Comment by iwishforducks** [comment]
Good stuff.

**Comment by deathrobotpunch** [comment]
I really like goonstation drop and roll where you rest and start alternating between A and D to put yourself out

File: code/datums/status_effects/buffs/stop_drop_roll.dm [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+	alert_type = null
+
+	tick_interval = 0.8 SECONDS
+
+/datum/status_effect/stop_drop_roll/on_apply()
+	if(!iscarbon(owner))
+		return FALSE
+
+	var/actual_interval = initial(tick_interval)
+	if(!owner.Knockdown(actual_interval * 2, ignore_canstun = TRUE) || owner.body_position != LYING_DOWN)
```
Comment (san7890) [inline code/datums/status_effects/buffs/stop_drop_roll.dm:12]: is `tick_interval` modified by the status_effect code? that's wack. not really a code concern as much as it is trying to learn what's goin on here

File: code/datums/status_effects/buffs/stop_drop_roll.dm [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+	RegisterSignal(owner, COMSIG_LIVING_SET_BODY_POSITION, PROC_REF(body_position_changed))
+
+	owner.visible_message(
+		span_danger("[owner] rolls on the floor, trying to put [owner.p_them()]self out!"),
+		span_notice("You stop, drop, and roll!"),
+	)
+	// Start with one weaker roll
+	owner.spin(spintime = actual_interval, speed = actual_interval / 4)
+	owner.adjust_fire_stacks(-0.5)
+	ADD_TRAIT(owner, TRAIT_HANDS_BLOCKED, id) // they're busy!
```
Comment (san7890) [inline code/datums/status_effects/buffs/stop_drop_roll.dm:26]: use `TRAIT_STATUS_EFFECT()` here?

File: code/datums/status_effects/buffs/stop_drop_roll.dm [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+	return TRUE
+
+/datum/status_effect/stop_drop_roll/on_remove()
+	UnregisterSignal(owner, list(COMSIG_MOVABLE_MOVED, COMSIG_LIVING_SET_BODY_POSITION))
+	REMOVE_TRAIT(owner, TRAIT_HANDS_BLOCKED, id)
```
Comment (san7890) [inline code/datums/status_effects/buffs/stop_drop_roll.dm:31]: ditto `TRAIT_STATUS_EFFECT()`

**Review by san7890 — CHANGES_REQUESTED** [review]
one nit that happens to be in two places

File: code/datums/status_effects/buffs/stop_drop_roll.dm (hunk 1 above)
Comment (MrMelbert) [reply code/datums/status_effects/buffs/stop_drop_roll.dm:12]: yeah, tick interval becomes `world.time + tick_interval` on creation

File: code/datums/status_effects/buffs/stop_drop_roll.dm (hunk 2 above)
Comment (MrMelbert) [reply code/datums/status_effects/buffs/stop_drop_roll.dm:26]: I opted not to so it'd just show up as `stop_drop_roll` in vv (for clarity), but if you think we'll add another stop drop and roll trait source down the line i'll change it

**Comment by MrMelbert** [comment]
Oh as an aside the way it's current set up, stop/drop/roll will cancel if someone drags you or space wind pushes you, not sure if we want that / care / if it's clunky. Thoughts?

File: code/datums/status_effects/buffs/stop_drop_roll.dm (hunk 2 above)
Comment (san7890) [reply code/datums/status_effects/buffs/stop_drop_roll.dm:26]: Ultimately meh then, just feels awk

**Comment by san7890** [comment]
I think it might be rather easy for people to throw an incendiary item at you and then abuse grabbing mechanics to prevent you from that motion... if there is a way to avoid that from happening I would like to see that (i.e. grabbing someone who is rolling is "thrashing too hard" to be grabbed onto)...

Could just be paranoia but I think it's easy to prolong firestack damage in close range combat through this. The best (balance) fix would be to have fire spread through contact when you grab someone but idk how the hell you'd code that (unless we have signals).

**Comment by SomeoneYouProbablyKnow** [comment]
> I think it might be rather easy for people to throw an incendiary item at you and then abuse grabbing mechanics to prevent you from that motion... if there is a way to avoid that from happening I would like to see that (i.e. grabbing someone who is rolling is "thrashing too hard" to be grabbed onto)...

As opposed to throwing the same item and laughing while you stun yourself for 5 whole seconds (Kinda why every second crewman is carrying a fire extinguisher, dam gear checks)

**Comment by san7890** [comment]
is this ready or do you plan to do more work on this? unsure due to the question you asked in your last post. if you wanna do more work on it please draft

**Comment by MrMelbert** [comment]
I guess we can merge it and change it later if it ends up being jank

**Comment by san7890** [comment]
i'm still sus on people to abuse that in the present combat meta but i'll eat the cost of that if it means better user qol

**Pull request merged**